File: 0515-find-largest-value-in-each-tree-row/0515-find-largest-value-in-each-tree-row.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    result = []
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        self.result = list()
        
        if not root:
            return self.result
        # self.result.append(/)
        self.helper(root, 1)

        return self.result
    
    def helper(self, root: Optional[TreeNode], level : int) -> None:
        #base
        if root is None:
            return
        #logic
        if level > len(self.result):
            self.result.append(root.val)
        
        if level <= len(self.result):
            if self.result[level - 1] < root.val:
                self.result[level - 1] = root.val
        
        self.helper(root.left, level + 1)
        self.helper(root.right, level + 1)

# DFS
#Time Complexity: O(n)
#Space Complexity: O(h)

        

File: 0515-find-largest-value-in-each-tree-row/0515-find-largest-value-in-each-tree-row.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> result;
    public List<Integer> largestValues(TreeNode root) {
        result = new ArrayList<>();

        if(root == null){
            return result;
        }

        helper(root, 1);

        return result;
    }

    private void helper(TreeNode root, Integer level){
        //base
        if(root == null){
            return;
        }

        if(level > result.size()){
            result.add(root.val);
        }

        if(level <= result.size()){
            if(result.get(level - 1) < root.val){
                result.set(level - 1, root.val);
            }            
        }

        helper(root.left, level + 1);
        helper(root.right, level + 1);
    }
}

// DFS
// Maintain level at each level and update the result at index when larger element is found
// Time Complexity: O(n)
// Space Complexity: O(h)
